From 3a8181557ce9b3cfe58fbdf8b1f3e7adc5c82ac7 Mon Sep 17 00:00:00 2001
From: Kathiravan Thirumoorthy <quic_kathirav@quicinc.com>
Date: Mon, 15 Apr 2024 18:09:53 +0530
Subject: [PATCH] coresight: tmc: handle the error cases properly

Probe of TMC ETR is deferred until the rproc device is available. In
such cases, we need to unregister the panic notifier, else system will
crash due to the UAF scenario.

Only for certain targets, we need to abort the TMC and write the markers
during SSR event. So when the qcom,rproc property is not available in
tmc node, skip the SSR handler registration, instead of returning the
error.

Change-Id: I7179573173b97113c184ca4ccd512a084ae3d827
Fixes: 2cba54300cd7 ("coresight: tmc-etr: add coresight ssr abort support")
Signed-off-by: Kathiravan Thirumoorthy <quic_kathirav@quicinc.com>
---
 drivers/hwtracing/coresight/coresight-tmc-core.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/hwtracing/coresight/coresight-tmc-core.c b/drivers/hwtracing/coresight/coresight-tmc-core.c
index c4eea17bc5c5..211bb4c95b48 100644
--- a/drivers/hwtracing/coresight/coresight-tmc-core.c
+++ b/drivers/hwtracing/coresight/coresight-tmc-core.c
@@ -552,10 +552,13 @@ static int tmc_probe(struct amba_device *adev, const struct amba_id *id)
 
 		if (of_property_read_u32(dev->of_node, "qcom,rproc",
 					 &rproc_node))
-			return -ENODEV;
+			goto skip_ssr;
+
 		rproc = rproc_get_by_phandle(rproc_node);
-		if (!rproc)
+		if (!rproc) {
+			atomic_notifier_chain_unregister(&panic_notifier_list, &drvdata->panic_blk);
 			return -EPROBE_DEFER;
+		}
 
 		drvdata->ssr_blk.notifier_call = tmc_etr_ssr_handler;
 		notifier = qcom_register_ssr_atomic_notifier(rproc->name, &drvdata->ssr_blk);
-- 
2.34.1

